### PR TITLE
fix: mark fresh install to cp the default configuration file directly

### PR DESCRIPTION
**Pull request by JimMoen**
Fixes [EMQX-12452](https://emqx.atlassian.net/browse/EMQX-12452) [EMQX-12347](https://emqx.atlassian.net/browse/EMQX-12347)

Release version: v/e5.7.1

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- ~[ ] Added tests for the changes~
- ~[ ] Added property-based tests for code which performs user input validation~
- ~[ ] Changed lines covered in coverage report~
- ~[ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files~
- ~[ ] For internal contributor: there is a jira ticket to track this change~
- ~[ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket~
- ~[ ] Schema changes are backward compatible~

## Checklist for CI (.github/workflows) changes

- ~[ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)~
- ~[ ] Change log has been added to `changes/` dir for user-facing artifacts update~

[EMQX-12452]: https://emqx.atlassian.net/browse/EMQX-12452?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[EMQX-12347]: https://emqx.atlassian.net/browse/EMQX-12347?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/emqx_management/src/emqx_mgmt_api_plugins.erl [hunk 1]
```diff
@@ -565,6 +564,7 @@ install_package(FileName, Bin) ->
     ok = filelib:ensure_dir(File),
     ok = file:write_file(File, Bin),
     PackageName = string:trim(FileName, trailing, ".tar.gz"),
+    put(?fresh_install, true),
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt_api_plugins.erl:567]: why not an argument passed to emqx_plugins:ensure_installed?

Comment (JimMoen) [reply apps/emqx_management/src/emqx_mgmt_api_plugins.erl:567]: Installation from the management API(`?fresh_install` here) and when a node joins a cluster will execute similar code paths, such as creating plugin configuration directories, building schhema serde, etc.

The only difference is `emqx_plugins:ensure_plugin_config/1`
Installation when joining a cluster requires obtaining the existing configuration of the cluster, and a new installation will directly copy the default configuration file from tarball.
Passing parameters to the emqx_plugins:ensure_installed function will result in redundant function clause and `case...of`

File: apps/emqx_plugins/src/emqx_plugins.erl [hunk 2]
```diff
@@ -664,12 +673,12 @@ do_ensure_started(NameVsn) ->
                 ok ->
                     Plugin = do_read_plugin(NameVsn),
                     ok = load_code_start_apps(NameVsn, Plugin);
-                {error, plugin_not_found} ->
+                {error, #{reason := Reason} = ReasonMap} ->
                     ?SLOG(error, #{
-                        error_msg => "plugin_not_found",
+                        error_msg => string_reason(Reason),
                         name_vsn => NameVsn
                     }),
```
Comment (zmstone) [inline apps/emqx_plugins/src/emqx_plugins.erl:680]: ?SLOG(error, #{msg => failed_to_start_plugin, name_vsn => NameVsn, reason => Reason}).

File: apps/emqx_plugins/src/emqx_plugins.erl [hunk 3]
```diff
@@ -1376,23 +1409,23 @@ prune_backup_files(Path) ->
...
-read_file_fun(Path, ErrMsg, #{read_mode := ?JSON_MAP}) ->
+read_file_fun(Path, Msg, #{read_mode := ?JSON_MAP}) ->
     fun() ->
         case hocon:load(Path, #{format => richmap}) of
             {ok, RichMap} ->
                 {ok, hocon_maps:ensure_plain(RichMap)};
             {error, Reason} ->
-                ErrMeta = #{error_msg => ErrMsg, reason => Reason},
+                ErrMeta = #{msg => Msg, reason => Reason},
                 throw(ErrMeta)
```
Comment (zmstone) [inline apps/emqx_plugins/src/emqx_plugins.erl:1429]:
```suggestion
                ErrMeta = #{reason => Reason, path => Path},
                throw(ErrMeta)
```

File: apps/emqx_management/src/emqx_mgmt_api_plugins.erl [hunk 4]
```diff
@@ -467,7 +466,7 @@ do_install_package(FileName, Bin) ->
             ),
             Reason =
                 case hd(Filtered) of
-                    {error, #{error_msg := Reason0}} -> Reason0;
+                    {error, #{msg := Reason0}} -> Reason0;
                     {error, #{reason := Reason0}} -> Reason0
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt_api_plugins.erl:470]: this Reason0 can be atom? and line 474 try to use it as iolist

File: apps/emqx_management/src/emqx_mgmt_api_plugins.erl [hunk 5]
```diff
@@ -625,9 +627,9 @@ do_update_plugin_config(NameVsn, AvroJsonMap, AvroValue) ->
 
 return(Code, ok) ->
     {Code};
-return(_, {error, #{error_msg := "bad_info_file", reason := {enoent, _} = Reason}}) ->
+return(_, {error, #{msg := "bad_info_file", reason := {enoent, _} = Reason}}) ->
     {404, #{code => 'NOT_FOUND', message => readable_error_msg(Reason)}};
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt_api_plugins.erl:631]:
```suggestion
return(_, {error, #{msg := "bad_info_file", reason := {enoent, Path} = Reason}}) ->
    {404, #{code => 'NOT_FOUND', message => iolist_to_binary([Path, " does not exist"])}};
```

File: apps/emqx_plugins/src/emqx_plugins.erl [hunk 6]
```diff
@@ -1376,23 +1409,23 @@ prune_backup_files(Path) ->
...
 
-read_file_fun(Path, ErrMsg, #{read_mode := ?RAW_BIN}) ->
+read_file_fun(Path, Msg, #{read_mode := ?RAW_BIN}) ->
     fun() ->
         case file:read_file(Path) of
             {ok, Bin} ->
                 {ok, Bin};
             {error, Reason} ->
-                ErrMeta = #{error_msg => ErrMsg, reason => Reason},
+                ErrMeta = #{msg => Msg, reason => Reason},
```
Comment (zmstone) [inline apps/emqx_plugins/src/emqx_plugins.erl:1418]: to unify the `{enoent, Path}` pattern.
```suggestion
                ErrMeta = #{msg => Msg, reason => {Reason, Path}},
```

File: apps/emqx_management/src/emqx_mgmt_api_plugins.erl [hunk 7]
```diff
@@ -411,7 +410,7 @@ upload_install(post, #{body := #{<<"plugin">> := Plugin}}) when is_map(Plugin) -
     %% File bin is too large, we use rpc:multicall instead of cluster_rpc:multicall
     NameVsn = string:trim(FileName, trailing, ".tar.gz"),
     case emqx_plugins:describe(NameVsn) of
-        {error, #{error_msg := "bad_info_file", reason := {enoent, _}}} ->
+        {error, #{msg := "bad_info_file", reason := {enoent, _}}} ->
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt_api_plugins.erl:413]:
```suggestion
        {error, #{msg := "bad_info_file", reason := {enoent, _Path}}} ->
```

File: apps/emqx_management/src/emqx_mgmt_api_plugins.erl [hunk 8]
```diff
@@ -625,9 +627,9 @@ do_update_plugin_config(NameVsn, AvroJsonMap, AvroValue) ->
 
 return(Code, ok) ->
     {Code};
-return(_, {error, #{error_msg := "bad_info_file", reason := {enoent, _} = Reason}}) ->
+return(_, {error, #{msg := "bad_info_file", reason := {enoent, _} = Reason}}) ->
     {404, #{code => 'NOT_FOUND', message => readable_error_msg(Reason)}};
-return(_, {error, #{error_msg := "bad_avro_config_file", reason := {enoent, _} = Reason}}) ->
+return(_, {error, #{msg := "bad_hocon_file", reason := {enoent, _} = Reason}}) ->
     {404, #{code => 'NOT_FOUND', message => readable_error_msg(Reason)}};
```
Comment (zmstone) [inline apps/emqx_management/src/emqx_mgmt_api_plugins.erl:633]:
```suggestion
return(_, {error, #{msg := "bad_hocon_file", reason := {enoent, Path} = Reason}}) ->
    {404, #{code => 'NOT_FOUND', message => iolist_to_binary[Path, " does not exist"])}};
```

File: apps/emqx_plugins/src/emqx_plugins.erl [hunk 9]
```diff
@@ -1376,23 +1409,23 @@ prune_backup_files(Path) ->
         Deletes
     ).
 
-read_file_fun(Path, ErrMsg, #{read_mode := ?RAW_BIN}) ->
+read_file_fun(Path, Msg, #{read_mode := ?RAW_BIN}) ->
```
Comment (zmstone) [inline apps/emqx_plugins/src/emqx_plugins.erl:1412]: it's better to keep `msg => "the_msg_string"` close to `?SLOG` macro.
I would recommend to simply `throw(Reason)` from this function, and keep Msg outside.

Comment (JimMoen) [reply apps/emqx_plugins/src/emqx_plugins.erl:1412]: will change it in refactor PR (in 5.8.0)

**Review by zmstone — APPROVED** [review]

**Pull request merged**